core/src/main/java/hudson/model/queue/WorkUnitContext.java

```diff
@@ -107,13 +109,32 @@ public WorkUnit getPrimaryWorkUnit() {
     /**
      * All the {@link Executor}s that jointly execute a {@link Task} call this method to synchronize on the start.
      */
+    @Restricted(NoExternalUse.class)
     public void synchronizeStart() throws InterruptedException {
-        startLatch.synchronize();
-        // the main thread will send a notification
         Executor e = Executor.currentExecutor();
+        if (e == null) {
+            throw new IllegalStateException("synchronizeStart shall only be called from an Executor");
+        }
         WorkUnit wu = e.getCurrentWorkUnit();
-        if (wu.isMainWork()) {
-            future.start.set(e.getCurrentExecutable());
+        if (wu == null) {
+            throw new IllegalStateException("synchronizeStart shall only be called after a WorkUnit has been assigned");
+        }
+        boolean mainWork = wu.isMainWork();
+        try {
+            startLatch.synchronize();
+            // the main thread will send a notification
+            if (mainWork) {
+                future.start.set(e.getCurrentExecutable());
+            }
+        } catch (Error | RuntimeException | InterruptedException problem) {
+            if (mainWork) {
+                future.start.set(problem);
+            }
+            throw problem;
+        } finally {
+            if (mainWork && !future.start.isDone()) {
+                future.start.set(e.getCurrentExecutable());
```

Comment:
Won't this one override the one set if an exception is raised?

Reply:
Yes, it overrides the same value. Anyway, I've changed the method code because we were at the same point, the new exceptions thrown when the _executable_ or the _current work uni_t is not set were keeping the `future` unset, so it may trigger the problem we were fixing after all.